tests/pytests/integration/ssh/test_state.py

```diff
@@ -561,3 +562,298 @@ def test_retcode_state_sls_id_render_exception(self, salt_ssh_cli):
     def test_retcode_state_top_run_fail(self, salt_ssh_cli):
         ret = salt_ssh_cli.run("state.top", "top.sls")
         assert ret.returncode == EX_AGGREGATE
+
+
+@pytest.fixture(scope="class")
+def pillar_tree_nested(base_env_pillar_tree_root_dir):
+    top_file = """
+    base:
+      'localhost':
+        - nested
+      '127.0.0.1':
+        - nested
+    """
+    nested_pillar = r"""
+    {%- do salt.log.warning("hithere: pillar was rendered") %}
+    monty: python
+    the_meaning:
+      of:
+        life: 42
+        bar: tender
+      for: what
+    """
+    top_tempfile = pytest.helpers.temp_file(
+        "top.sls", top_file, base_env_pillar_tree_root_dir
+    )
+    nested_tempfile = pytest.helpers.temp_file(
+        "nested.sls", nested_pillar, base_env_pillar_tree_root_dir
+    )
+    with top_tempfile, nested_tempfile:
+        yield
+
+
+@pytest.mark.usefixtures("pillar_tree_nested")
+def test_pillar_is_only_rendered_once_without_overrides(salt_ssh_cli, caplog):
+    ret = salt_ssh_cli.run("state.apply", "test")
+    assert ret.returncode == 0
+    assert isinstance(ret.data, dict)
+    assert ret.data
+    assert ret.data[next(iter(ret.data))]["result"] is True
+    assert caplog.text.count("hithere: pillar was rendered") == 1
+
+
+@pytest.mark.usefixtures("pillar_tree_nested")
+def test_pillar_is_rerendered_with_overrides(salt_ssh_cli, caplog):
+    ret = salt_ssh_cli.run("state.apply", "test", pillar={"foo": "bar"})
+    assert ret.returncode == 0
+    assert isinstance(ret.data, dict)
+    assert ret.data
+    assert ret.data[next(iter(ret.data))]["result"] is True
+    assert caplog.text.count("hithere: pillar was rendered") == 2
+
+
+@pytest.mark.slow_test
+@pytest.mark.usefixtures("pillar_tree_nested")
+class TestStatePillarOverride:
```

Comment:
```suggestion
@pytest.mark.slow_test
@pytest.mark.usefixtures("pillar_tree_nested")
class TestStatePillarOverride:
```
Hmm, could we instead of classes split these tests into another file?
Something like:
```
tests/pytests/integration/ssh/state/test_state.py
tests/pytests/integration/ssh/state/test_pillar_override.py
tests/pytests/integration/ssh/state/test_pillar_override_template.py
```
etc ...

Reply:
Sure. That will result in a couple of merge conflicts on my other salt-ssh PRs (https://github.com/saltstack/salt/pull/64542 https://github.com/saltstack/salt/pull/65517), for which I just got the CI passing without failed flaky tests. ^^ Thinking about it, that probably would have happened anyways. Coming up.

Reply:
Refactoring is done, but I should probably refactor this again like you showed in https://github.com/saltstack/salt/pull/64542 (saw it after I re-requested your review here...)

Reply:
Will mark this as resolved since https://github.com/saltstack/salt/pull/65484/commits/0e26e667f3b7a6e239c69d7dc9bcb34d75e981b2 implemented this and https://github.com/saltstack/salt/pull/65484/commits/1eb86478e1d27ff00df3a347ef86992c1a3d6ca7 additionally refactored the tests into a single parametrized test function where sensible.